zyte_api/aio/client.py

```diff
@@ -43,6 +51,38 @@ def _post_func(session):
         return session.post
 
 
+class ExtractResult(Mapping):
+    """Result of a call to AsyncClient.extract.
+
+    It can be used as a dictionary to access the raw API response.
+
+    It also provides some helper properties for easier access to some of its
+    underlying data.
+    """
+
+    def __init__(self, api_response: dict):
+        self._api_response = api_response
+        self._http_response_body: Union[bytes|_NotLoaded] = _NOT_LOADED
+
+    def __getitem__(self, key):
+        return self._api_response[key]
+
+    def __iter__(self):
+        yield from self._api_response
+
+    def __len__(self):
+        return len(self._api_response)
+
+    @property
+    def http_response_body(self) -> Union[bytes|_NotLoaded]:
```

Comment:
It seems there could be a big overlap with data structures defined in https://github.com/scrapinghub/web-poet/blob/master/web_poet/page_inputs/http.py

Reply:
+1

I'm thinking if `python-zyte-api` should somehow import `web-poet` to utilize such page_inputs or should the page_inputs themselves be moved outside of `web-poet`. 

Later on, we'd need the functionalities of `web_poet.page_inputs.http` to process content-encoding when dealing with `browserHtml`.

Reply:
I’ll see you and raise you: https://github.com/scrapy-plugins/scrapy-zyte-api/pull/10 may be in a similar situation.

Reply:
Do we want `ExtractResult` here to be a subclass of web_poet’s HttpResponse, or expose a compatible interface? Or is the problem only not to reinvent the wheel code-wise? Or do we want `ExtractResult` to expose `httpResponseBody`, `httpResponseHeaders` and `browserHtml` through 2 HttpResponse objects, or possibly a different, new `web_poet` object in the case of `browserHtml`?

Reply:
> Do we want ExtractResult here to be a subclass of web_poet’s HttpResponse, or expose a compatible interface? Or is the problem only not to reinvent the wheel code-wise?

I believe it's both. We'd want the ecosystem revolving around Zyte's extraction and crawling be similar in interface (i.e. `web-poet`, `scrapy-poet`, `python-zyte-api`, `scrapy-zyte-api`, etc). 

I'm trying to think if there are some downsides in using `web-poet` for the Zyte API's client, but I can't think of any. I think it's more beneficial since `web-poet` would be used behind Zyte API. This means both server and client would use the same dependency promoting overall compatibility.

> Or do we want ExtractResult to expose httpResponseBody, httpResponseHeaders and browserHtml through 2 HttpResponse objects, or possibly a different, new web_poet object in the case of browserHtml?

After trying to make `httpResponseBody` work with Text Responses in https://github.com/scrapy-plugins/scrapy-zyte-api/pull/10, I realized that we'll need to read the headers easily. `web_poet.httpResponseHeaders` could easily accommodate Zyte API's formatting. For example, if we'd want to determine if a given response is text-based or not, we can check out the `Content-Type` header. However, Zyte API returns the headers like this:

```python
"httpResponseHeaders": [..., {'name': 'content-type', 'value': 'text/html; charset=UTF-8'}, ...]
```

To search for this value, we'd need to iterate through the list of header key-value pairs. On the other hand, `web-poet` would easily have this as:

```python
>>> headers = HttpResponseHeaders.from_name_value_pairs(zyte_api_response["httpResponseHeaders"])
>>> headers.get('Content-Type')
'text/html; charset=UTF-8'
```